improve docs for connecting Pipelines SDK to Kubeflow

thesuperzapper opened the pull request:
Closes https://github.com/kubeflow/website/issues/2916

__This PR continues the great work done by https://github.com/kubeflow/website/pull/2905:__
- adds an example for "out of cluster" dex multi-user:
   - resolves https://github.com/kubeflow/website/issues/2916
- creates more visual separation between sections of the page:
   - adds `---` separators between major sections
   - uses more H2 / H3 (rather than H4 / H5)
- re-words some of the more confusing sentences
- renames `Connect to Kubeflow Pipelines from the same cluster` to `Connect to Kubeflow Pipelines from within the same cluster`

Bobgy commented:
Thank you! I will take a look when I have time 🙂

zijianjoy commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+For Kubeflow Pipelines in non-multi-user mode, you can expose the Pipelines API locally by using `kubectl port-forward`.
 
-Kubeflow Pipelines standalone deploys a Kubernetes service named `ml-pipeline-ui` in your Kubernetes cluster without extra authentication.
+Kubeflow Pipelines (standalone) deploys an unsecured Kubernetes `Service` named `ml-pipeline-ui` which requires no authentication.
```
Suggestion:

`Kubeflow Pipelines Standalone doesn't enforce authentication in `ml-pipeline-ui` service`.

thesuperzapper replied:
Changed.

zijianjoy commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
 import kfp
-namespace = 'kubeflow' # or the namespace you deployed Kubeflow Pipelines
+
+namespace = 'kubeflow' # or the namespace in which you deployed Kubeflow
```
To be explicit about what is deployed in this namespace:

```
# or the namespace in which you deployed Kubeflow Pipelines
```

thesuperzapper replied:
Changed.

Bobgy reviewed (commented):
Thank you!
I love a lot of the changes you are making to make the doc more readable!
Here are some quick nit pickings.

I'd like to separately discuss overall structure first:

Have you considered using multi-user mode vs non-multiuser mode as the top level sections? and then in each section, dive into how to connect from outside/inside the cluster.

I am considering this, because multi-user mode auth/authz is a beast, it's always complex. I think starting with the simpler one is better for user knowledge. Similarly, we can easily introduce why multi-user mode is complex in that big section.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
-The SDK client can send requests to this endpoint to upload pipelines, create pipeline runs, schedule recurring runs, and more.
-
+* Deploy Kubeflow Pipelines using one of the [installation options](/docs/components/pipelines/installation/overview/)
+* [Install the Kubeflow Pipelines SDK](/docs/components/pipelines/sdk/install-sdk/) python library
```
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+### Non-multi-user mode
 
-For [Kubeflow Pipelines standalone](https://www.kubeflow.org/docs/components/pipelines/installation/standalone-deployment/) and [Google Cloud AI Platform Pipelines](/docs/components/pipelines/installation/overview/#google-cloud-ai-platform-pipelines), you can also connect to the API via `kubectl port-forward`.
+For Kubeflow Pipelines in non-multi-user mode, you can expose the Pipelines API locally by using `kubectl port-forward`.
```
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
 
-Note, for Kubeflow Pipelines in multi-user mode, you cannot access the API using kubectl port-forward
-because it requires authentication. Refer to distribution specific documentation as recommended above.
+### Multi-User mode
```
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+import requests
+from urllib.parse import urlsplit
+
+def get_istio_auth_session(url: str, username: str, password: str) -> dict:
```
I don't really know of a better place to put this.

Also, since dex is used in the default `kubeflow/manifests`, it probably doesn't hurt to include it here.

thesuperzapper replied:
This was also discussed in https://github.com/kubeflow/website/issues/2916 where @Bobgy agreed (https://github.com/kubeflow/website/issues/2916#issuecomment-914265013) it makes sense to include in the base pipelines docs.

Because Dex is used by so many Kubeflow distributions.

zijianjoy replied:
Sounds good on having the code snippet in the doc then. Hope that it doesn't create too much difficulty in keeping the code snippet up-to-date.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+---
+
+
+## Connect to Kubeflow Pipelines from within the same cluster
```
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
````diff
@@ -233,17 +343,18 @@ Or configure in a Jupyter Notebook by using the [IPython built-in `%env` magic c
 ```
 
-Refer to [more configurable environment variables here](https://github.com/kubeflow/pipelines/blob/54ac9a6a7173aecbbb30a043b2077e790cac6953/sdk/python/kfp/_client.py#L84-L90).
+Find the other Kubeflow Pipelines SDK environment variables [here](https://github.com/kubeflow/pipelines/blob/991473d3e42eb64a9ec5a7fc44a6733d1bd1db70/sdk/python/kfp/_client.py#L93-L99).
````
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+
+A Kubernetes [ServiceAccount token](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) volume can be mounted to the Pod and used by the Kubeflow Pipelines SDK.
+
+Simply put, a ServiceAccount token is injected into the Pod (by Kubernetes), then the Kubeflow Pipelines SDK uses this token to identity itself to the Kubeflow Pipelines API.
```
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+- you must use `pipelines.kubeflow.org` as the token's `audience`
+- no additional setup is required to refresh tokens, refreshing is managed by Kubernetes itself
+
+#### Access Kubeflow Pipelines from Jupyter notebook
```
Changed.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+subjects:
+  - kind: ServiceAccount
+    name: default-editor
+    namespace: namespace-1
```
Thanks, I was wrong, I have fixed it.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+  - image: hello-world:latest
+    name: hello-world
+    env:
+      - name: KF_PIPELINES_SA_TOKEN_PATH
+        value: /var/run/secrets/kubeflow/pipelines/token
```
I think it is useful for users to be able to see the config, even if it's default.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+
+Simply put, a ServiceAccount token is injected into the Pod (by Kubernetes), then the Kubeflow Pipelines SDK uses this token to identity itself to the Kubeflow Pipelines API.
+
+Note, the Kubeflow Pipelines API understands Kubernetes RBAC, and will respect `RoleBindings` assigned to the ServiceAccount.
```
Do you have a place where we can link to?

zijianjoy replied:
Maybe this link? https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/view-edit-cluster-roles.yaml#L43-L97

thesuperzapper replied:
Added in https://github.com/kubeflow/website/pull/3021/commits/7e2805390335072cc33691ed80e1e7a4c3245256

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -4,227 +4,337 @@ description = "How to connect to Kubeflow Pipelines using the SDK client and con
+* [IBM Cloud Kubernetes Service (IKS)](/docs/distributions/ibm/pipelines/)
 
-### Non-multi-user mode
+__Example for deployments that use [dex](https://dexidp.io/).__
```
Changed.

thesuperzapper commented:
@Bobgy @zijianjoy sorry for the long delay, but I have rebased and applied the requested changes.

Hopefully its ready to merge now!

thesuperzapper commented:
@Bobgy this PR is still waiting to be merged.

All the significant comments have been addressed, and I think its important for users to have this doc updated.

### Screenshot of new page

![Connect-to-the-Kubeflow-Pipelines-API-Kubeflow](https://user-images.githubusercontent.com/5735406/143530533-468b1e36-a200-434e-b694-08292c2eb7dc.png)

zijianjoy commented:
/lgtm

Thank you Matthew!

thesuperzapper commented:
@zijianjoy I have added the link you suggested in https://github.com/kubeflow/website/pull/3021#discussion_r757265380.

This PR should be ready to merge now if you `/lgtm` and `/approve`

zijianjoy commented:
/approve

Thank you Mathew!

thesuperzapper commented:
@zijianjoy sorry I was still not happy with these docs, so I have worked on them a bit, and now I think they are MUCH easier to read, do you mind giving it a quick look?

Here are screenshots of the two changed pages.

## Connect the Pipelines SDK to Kubeflow Pipelines
![Connect-the-Pipelines-SDK-to-Kubeflow-Pipelines-Kubeflow](https://user-images.githubusercontent.com/5735406/145762352-dd6c3b8c-acaf-4780-8500-552863474143.png)

## Multi-user Isolation
![Multi-user-Isolation-Kubeflow](https://user-images.githubusercontent.com/5735406/145762386-7f6b7e49-d257-437a-bfa5-852a0b5e6a32.png)

zijianjoy commented on content/en/docs/components/pipelines/overview/multi-user.md:
```diff
@@ -1,175 +1,179 @@
-* For Google Cloud: [In-cluster authentication to Google Cloud from Kubeflow](/docs/gke/authentication/#in-cluster-authentication)
+{{% alert title="Warning" color="warning" %}}
+Kubeflow makes no security guarantees about profile isolation.
+It is very likely that malicious attempts to infiltrate other user's profiles will succeed.
```
It seems not necessary to include this line, the first line is sufficient to deliver the message of no security guarantee.

jbottum replied:
I agree. We should consider removing or at least rewording.

thesuperzapper replied:
Resolved in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

zijianjoy commented on content/en/docs/components/pipelines/overview/multi-user.md:
```diff
@@ -1,175 +1,179 @@
 
-* [Grant user minimal Kubernetes cluster access](/docs/components/multi-tenancy/getting-started/#pre-requisites-grant-user-minimal-kubernetes-cluster-access)
-* [Managing contributors through the Kubeflow UI](/docs/components/multi-tenancy/getting-started/#managing-contributors-through-the-kubeflow-ui)
-* For Google Cloud: [In-cluster authentication to Google Cloud from Kubeflow](/docs/gke/authentication/#in-cluster-authentication)
```
This is useful to keep this hyperlink for helping user to find right set of authentication method.

thesuperzapper replied:
This diff is misleading, the links are actually still present (except the google cloud one).

zijianjoy commented on content/en/docs/components/pipelines/overview/multi-user.md:
```diff
@@ -1,175 +1,179 @@
+It is very likely that malicious attempts to infiltrate other user's profiles will succeed.
+{{% /alert %}}
 
-Note, Kubeflow Pipelines multi-user isolation is only supported in
-[the full Kubeflow deployment](/docs/components/pipelines/installation/overview/#full-kubeflow-deployment)
-starting from Kubeflow v1.1 and **currently** on all platforms except OpenShift. For the latest status about platform support, refer to [kubeflow/manifests#1364](https://github.com/kubeflow/manifests/issues/1364#issuecomment-668415871).
```
Why do we remove this section?

thesuperzapper replied:
Re-added in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

zijianjoy commented on content/en/docs/components/pipelines/overview/multi-user.md:
```diff
@@ -1,175 +1,179 @@
+
+# when `path` is None, the `KF_PIPELINES_SA_TOKEN_PATH` environment variable is used
+# (if `KF_PIPELINES_SA_TOKEN_PATH` is unset, the default is: `/var/run/secrets/kubeflow/pipelines/token`)
+credentials = kfp.auth.ServiceAccountTokenVolumeCredentials(path=None)
```
Is this step necessary? Or we are listing here in case people want to use it?

thesuperzapper replied:
It's only explicitly necessary if the token is not mounted at the default location, but since we have no idea how the user's environment looks, I think it's best to be explicit.

zijianjoy commented on content/en/docs/components/pipelines/overview/multi-user.md:
```diff
@@ -1,175 +1,179 @@
 )
+print("-------- BEGIN: RUNS --------")
 print(runs)
+print("-------- END: RUN --------")
```
NIT: `END: RUNS` instead of `END: RUNS`

thesuperzapper replied:
Resolved in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

bartgras commented:
@zijianjoy @thesuperzapper

It's such an improvement comparing to the current version of the documentation. Even merging it without further fixes would be very beneficial to the community.
These small nitpicks could be addressed in separate PR. I could help with that.

thesuperzapper commented:
@bartgras I agree, I will reopen this PR

/reopen

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -1,255 +1,359 @@
-## Before you begin
+In [standalone deployments](/docs/components/pipelines/installation/standalone-deployment/) of Kubeflow Pipelines, there is no authentication enforced on the `ml-pipeline-ui` service.
 
-* You need a Kubeflow Pipelines deployment using one of the [installation options](/docs/components/pipelines/installation/overview/).
-* [Install the Kubeflow Pipelines SDK](/docs/components/pipelines/sdk/install-sdk/).
```
Re-added in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -1,255 +1,359 @@
 
+{{% alert title="Warning" color="warning" %}}
+This will only work for standalone Kubeflow Pipelines, for multi-user Kubeflow see [Multi-user Kubeflow (from inside cluster)](#multi-user-kubeflow-from-inside-cluster).
+{{% /alert %}}
```
It's for visual clarity, and to prevent people blindly copying the code in the "standalone" sections and it not working in a "full" kubeflow deployment.

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -1,255 +1,359 @@
 +++
 
-This guide demonstrates how to connect to Kubeflow Pipelines using [the Kubeflow Pipelines SDK client](/docs/components/pipelines/sdk/sdk-overview/), and how to [configure the SDK client using environment variables](#configure-sdk-client-by-environment-variables).
+How you connect the Pipelines SDK to Kubeflow Pipelines will depend on what type of Kubeflow deployment you have, and where you are running your code:
```
Resolved in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -1,255 +1,359 @@
 +++
 
-This guide demonstrates how to connect to Kubeflow Pipelines using [the Kubeflow Pipelines SDK client](/docs/components/pipelines/sdk/sdk-overview/), and how to [configure the SDK client using environment variables](#configure-sdk-client-by-environment-variables).
+How you connect the Pipelines SDK to Kubeflow Pipelines will depend on what type of Kubeflow deployment you have, and where you are running your code:
```
Resolved in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

thesuperzapper commented on content/en/docs/components/pipelines/sdk/connect-api.md:
```diff
@@ -1,255 +1,359 @@
-This guide demonstrates how to connect to Kubeflow Pipelines using [the Kubeflow Pipelines SDK client](/docs/components/pipelines/sdk/sdk-overview/), and how to [configure the SDK client using environment variables](#configure-sdk-client-by-environment-variables).
+How you connect the Pipelines SDK to Kubeflow Pipelines will depend on what type of Kubeflow deployment you have, and where you are running your code:
 
+* [Standalone Kubeflow (from inside cluster)](#standalone-kubeflow-from-inside-cluster)
```
Resolved in https://github.com/kubeflow/website/pull/3021/commits/c3bf6da6cbc8e517174b1d91cdbf9e034464eb6e

thesuperzapper commented:
@zijianjoy after MUCH delay, I have addressed the review comments, and I think this PR is ready to merge (after a quick review from someone such as yourself)!

_NOTE: I have also added "collapsible" sections on " Connect the Pipelines SDK to Kubeflow Pipelines" because the page is far too long to browse comfortably otherwise._

__Here are some screenshots from the latest changes:__

### Connect the Pipelines SDK to Kubeflow Pipelines

<details>
<summary>Click to View (collapsed view)</summary>

![Connect-the-Pipelines-SDK-to-Kubeflow-Pipelines-Kubeflow](https://user-images.githubusercontent.com/5735406/188113783-d6c68f81-d2c1-41da-be8d-fdf8d859a815.png)

</details>

<details>
<summary>Click to View (fully expanded view)</summary>

![Connect-the-Pipelines-SDK-to-Kubeflow-Pipelines-Kubeflow (expanded)](https://user-images.githubusercontent.com/5735406/188113947-746baf28-db43-468d-84d7-6dfdd755584a.png)

</details>

### Multi-user Isolation

<details>
<summary>Click to View</summary>

![Multi-user-Isolation-Kubeflow](https://user-images.githubusercontent.com/5735406/188114018-b75ab1e8-a78a-49e1-91bb-cf4a1224098b.png)

</details>

thesuperzapper commented:
@zijianjoy I would greatly appreciate a quick look over so we can get this significant docs improvement merged. 

I am fairly sure its ready to merge.

thesuperzapper commented:
@connor-mccarthy @chensun I see that you are actively working on the KFP docs, would you have time to review this PR so it can be merged? (especially so I don't have to keep rebasing)

Here are links to the two changed pages:
- https://deploy-preview-3021--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/v1/overview/multi-user/
- https://deploy-preview-3021--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/v1/sdk/connect-api/

connor-mccarthy commented on content/en/docs/components/pipelines/v1/overview/multi-user.md:
```diff
@@ -1,175 +1,178 @@
-    )],
-))
+import kfp
+from kfp_server_api import *
```
nit: Why `import *`? This is generally a discouraged import pattern.

thesuperzapper replied:
@connor-mccarthy I agree, I just left it the same as it was, feel free to raise a quick PR to fix it?

connor-mccarthy reviewed (approved):
> It's such an improvement comparing to the current version of the documentation. Even merging it without further fixes would be very beneficial to the community.

Agreed.

This is a great improvement. Note: this has been reviewed several times over for content and, since this is largely a stylistic and clarity improvement, I have not tested the content.

/lgtm
/approve
